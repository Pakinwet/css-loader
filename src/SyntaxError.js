import formatCodeFrame from 'babel-code-frame';

export default class SyntaxError extends Error {
  constructor(error) {
    super(error);

    const { reason, message, line, column, source } = error;

    this.name = 'SyntaxError';
    this.message = reason || message;

    if (line && column) {
      this.message += ` (${line}:${column})`;

      if (source) {
        this.message += `\n\n${formatCodeFrame(source, line, column)}\n`;
      }
    }

    Error.captureStackTrace(this, this.constructor);
  }
}
