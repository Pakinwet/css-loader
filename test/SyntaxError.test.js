import SyntaxError from '../src/SyntaxError';

import webpack from './helpers/compiler';
import { normalizeErrors } from './helpers/utils';

describe('SyntaxError', () => {
  test('basic', () => {
    expect(new SyntaxError(new Error()).toString()).toMatchSnapshot();
    expect(new SyntaxError(new Error('message')).toString()).toMatchSnapshot();
    expect(
      new SyntaxError(new Error('message', 'test.css')).toString()
    ).toMatchSnapshot();
    expect(
      new SyntaxError(new Error('message', 'test.css', 100)).toString()
    ).toMatchSnapshot();

    const errorWithReasong = new Error();

    errorWithReasong.reason = 'reason';

    expect(new SyntaxError(errorWithReasong).toString()).toMatchSnapshot();

    const errorWithLineAndColumn = new Error('message');

    errorWithLineAndColumn.line = 1;
    errorWithLineAndColumn.column = 1;

    expect(
      new SyntaxError(errorWithLineAndColumn).toString()
    ).toMatchSnapshot();

    const errorWithLineAndColumnAndSource = new Error('message');

    errorWithLineAndColumnAndSource.line = 1;
    errorWithLineAndColumnAndSource.column = 1;
    errorWithLineAndColumnAndSource.source = '.class { invalid }';

    expect(
      new SyntaxError(errorWithLineAndColumnAndSource).toString()
    ).toMatchSnapshot();
  });

  test('integration', async () => {
    const stats = await webpack('broken.js');

    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });
});
