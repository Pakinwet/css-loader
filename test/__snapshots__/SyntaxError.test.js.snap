// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyntaxError basic 1`] = `"SyntaxError"`;

exports[`SyntaxError basic 2`] = `"SyntaxError: message"`;

exports[`SyntaxError basic 3`] = `"SyntaxError: message"`;

exports[`SyntaxError basic 4`] = `"SyntaxError: message"`;

exports[`SyntaxError basic 5`] = `"SyntaxError: reason"`;

exports[`SyntaxError basic 6`] = `"SyntaxError: message (1:1)"`;

exports[`SyntaxError basic 7`] = `
"SyntaxError: message (1:1)

> 1 | .class { invalid }
    | ^
"
`;

exports[`SyntaxError integration: errors 1`] = `
Array [
  [ModuleBuildError: Module build failed (from \`replaced original path\`):
SyntaxError: Unknown word (2:3)

  1 | .some {
> 2 |   invalid css;
    |   ^
  3 | }
  4 | 
],
]
`;

exports[`SyntaxError integration: warnings 1`] = `Array []`;
